Sources/NIOSOCKS/Channel Handlers/SOCKSClientHandler.swift

```diff
@@ -0,0 +1,118 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2021 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+
+/// Data was unexpectedly written or read before the SOCKS proxy
+/// connection has been fully established.
+public struct ProxyNotEstablished: Error {
+    
+}
+
+/// Connects to a SOCKS server to establish a proxied connection
+/// to a host. This handler should be inserted at the beginning of a
+/// channel's pipeline.
+public class SOCKSClientHandler: ChannelDuplexHandler {
+    
+    public typealias InboundIn = ByteBuffer
+    public typealias InboundOut = ByteBuffer
+    public typealias OutboundIn = ByteBuffer
+    public typealias OutboundOut = ByteBuffer
+    
+    public let supportedAuthenticationMethods: [AuthenticationMethod]
+    public let targetAddress: AddressType
+    public let targetPort: UInt16
+    
+    private var state: ClientStateMachine
+    private var buffered: ByteBuffer
+    
+    public init(supportedAuthenticationMethods: [AuthenticationMethod], targetAddress: AddressType, targetPort: UInt16) {
+        precondition(supportedAuthenticationMethods.count <= 255)
+        self.supportedAuthenticationMethods = supportedAuthenticationMethods
+        self.state = ClientStateMachine()
+        self.buffered = ByteBuffer()
+        self.targetAddress = targetAddress
+        self.targetPort = targetPort
+    }
+    
+    public func channelInactive(context: ChannelHandlerContext) {
+        context.fireChannelInactive()
+    }
+    
+    public func channelActive(context: ChannelHandlerContext) {
+        let greeting = ClientGreeting(
+            methods: self.supportedAuthenticationMethods
+        )
+        var buffer = ByteBuffer()
+        buffer.writeClientGreeting(greeting)
+        self.state.sendClientGreeting(greeting)
+        
+        context.writeAndFlush(self.wrapOutboundOut(buffer), promise: nil)
```

Comment:
This initial logic should be factored out into a method that can be called from either here or `handlerAdded` if `channel.isActive`. The reason is that we can be added late, and we want to support that as well.

Additionally, we should use `self.state` to guard against that method being invoked twice.